File: client/src/models/bloodRequest.ts
// Mirrors lib/models/blood_request.dart (BloodUrgency + BloodRequest)
// and lib/services/donor_profile_service.dart (DonorProfile)
export type BloodUrgencyName = 'routine' | 'urgent' | 'critical';

export interface BloodUrgency {
  name: BloodUrgencyName;
  label: string;
}

export const BLOOD_URGENCIES: BloodUrgency[] = [
  { name: 'routine', label: 'Routine' },
  { name: 'urgent', label: 'Urgent' },
  { name: 'critical', label: 'Critical' },
];

export function urgencyLabel(name: BloodUrgencyName): string {
  const found = BLOOD_URGENCIES.find((u) => u.name === name);
  return found ? found.label : name;
}

export interface BloodRequest {
  id: string;
  patientName: string;
  bloodGroup: string;
  hospital: string;
  location: string;
  units: number;
  urgency: BloodUrgencyName;
  requiredBy: Date;
  contactNumber: string;
  notes: string;
}

/**
 * Carries the SERVER's eligibility verdict rather than deriving one — the
 * 90-day donation interval is a medical rule, evaluated once in the API so
 * every client renders the same answer. See utils/bloodEligibility.ts, which
 * now holds only display formatting, not the rule itself.
 */
export interface DonorProfile {
  available: boolean;
  bloodGroup?: string;
  lastDonated?: Date;
  eligible: boolean;
  daysUntilEligible: number;
  progress: number;
  statusCopy: string;
}

export const EMPTY_DONOR_PROFILE: DonorProfile = {
  available: false,
  bloodGroup: undefined,
  lastDonated: undefined,
  eligible: true,
  daysUntilEligible: 0,
  progress: 1,
  statusCopy: 'No donation recorded yet',
};

// Two-letter initials used in the request card avatar (BloodRequest.initials)
export function requestInitials(patientName: string): string {
  const parts = patientName.trim().split(/\s+/).filter(Boolean);
  if (parts.length === 0) return '?';
  if (parts.length === 1) return parts[0].substring(0, 1).toUpperCase();
  return (
    parts[0].substring(0, 1) + parts[parts.length - 1].substring(0, 1)
  ).toUpperCase();
}

// Short date formatter shared across the blood-bank screens (formatShortDate)
export function formatShortDate(d: Date): string {
  const months = [
    'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
    'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
  ];
  return `${d.getDate().toString().padStart(2, '0')} ${months[d.getMonth()]} ${d.getFullYear()}`;
}

File: mobile/aust-central/lib/data/models/blood_request.dart
import 'package:flutter/material.dart';

enum BloodUrgency {
  routine,
  urgent,
  critical;

  static BloodUrgency fromName(String? raw) {
    switch (raw) {
      case 'urgent':
        return BloodUrgency.urgent;
      case 'critical':
        return BloodUrgency.critical;
      case 'routine':
      default:
        return BloodUrgency.routine;
    }
  }

  String get label {
    switch (this) {
      case BloodUrgency.routine:
        return 'Routine';
      case BloodUrgency.urgent:
        return 'Urgent';
      case BloodUrgency.critical:
        return 'Critical';
    }
  }
}

@immutable
class BloodRequest {
  final String id;
  final String patientName;
  final String bloodGroup;
  final String hospital;
  final String location;
  final int units;
  final BloodUrgency urgency;
  final DateTime requiredBy;
  final String contactNumber;
  final String notes;

  const BloodRequest({
    required this.id,
    required this.patientName,
    required this.bloodGroup,
    required this.hospital,
    required this.location,
    required this.units,
    required this.urgency,
    required this.requiredBy,
    required this.contactNumber,
    this.notes = '',
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'patientName': patientName,
        'bloodGroup': bloodGroup,
        'hospital': hospital,
        'location': location,
        'units': units,
        'urgency': urgency.name,
        'requiredBy': requiredBy.toIso8601String(),
        'contactNumber': contactNumber,
        'notes': notes,
      };

  factory BloodRequest.fromJson(Map<String, dynamic> json) => BloodRequest(
        id: json['id'] as String,
        patientName: json['patientName'] as String,
        bloodGroup: json['bloodGroup'] as String,
        hospital: json['hospital'] as String,
        location: json['location'] as String? ?? '',
        units: (json['units'] as num).toInt(),
        urgency: BloodUrgency.fromName(json['urgency'] as String?),
        requiredBy: DateTime.parse(json['requiredBy'] as String),
        contactNumber: json['contactNumber'] as String,
        notes: json['notes'] as String? ?? '',
      );

  /// Two-letter initials used in the request card avatar.
  String get initials {
    final parts =
        patientName.trim().split(RegExp(r'\s+')).where((p) => p.isNotEmpty);
    if (parts.isEmpty) return '?';
    if (parts.length == 1) return parts.first.substring(0, 1).toUpperCase();
    return (parts.first.substring(0, 1) + parts.last.substring(0, 1))
        .toUpperCase();
  }
}
